overlord/servicestate/quota_handlers.go

```diff
@@ -37,6 +40,302 @@ import (
 	"github.com/snapcore/snapd/wrappers"
 )
 
+// QuotaControlAction is the serialized representation of a quota group
+// modification that lives in a task.
+type QuotaControlAction struct {
+	// QuotaName is the name of the quota group being controlled.
+	QuotaName string `json:"quota-name"`
+
+	// Action is the action being taken on the quota group. It can be either
+	// "create", "update", or "remove".
+	Action string `json:"action"`
+
+	// AddSnaps is the set of snaps to add to the quota group, valid for either
+	// the "update" or the "create" actions.
+	AddSnaps []string `json:"snaps"`
+
+	// MemoryLimit is the memory limit for the quota group being controlled,
+	// either the initial limit the group is created with for the "create"
+	// action, or if non-zero for the "update" the memory limit, then the new
+	// value to be set.
+	MemoryLimit quantity.Size
+
+	// ParentName is the name of the parent for the quota group if it is being
+	// created. Eventually this could be used with the "update" action to
+	// support moving quota groups from one parent to another, but that is
+	// currently not supported.
+	ParentName string
+}
+
+func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	meter := snapstate.NewTaskProgressAdapterUnlocked(t)
+
+	qcs := []QuotaControlAction{}
+	err := t.Get("quota-control-actions", &qcs)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get quota-control-action: %v", err)
+	}
+
+	// TODO: support more than one action
+	switch {
+	case len(qcs) > 1:
+		return fmt.Errorf("multiple quota group actions not supported yet")
+	case len(qcs) == 0:
+		return fmt.Errorf("internal error: no quota group actions for quota-control task")
+	}
+
+	qc := qcs[0]
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	args := quotaControlArgs{
+		St:          st,
+		AllGrps:     allGrps,
+		Action:      &qc,
+		Meter:       meter,
+		PerfTimings: perfTimings,
+	}
+
+	switch qc.Action {
+	case "create":
+		err = doCreateQuota(args)
+	case "remove":
+		err = doRemoveQuota(args)
+	case "update":
+		err = doUpdateQuota(args)
+	default:
+		err = fmt.Errorf("unknown action %q requested", qc.Action)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	t.SetStatus(state.DoneStatus)
```

Comment:
mmh, because ensureSnapServicesForGroup itself does Unlock the state this is too late :(, we either need to make the operations really idempotent (with tests) or we need to restructure ensureSnapServicesForGroup a bit and place this carefully inside it. I'm happy to discuss this further. Btw we probably want to extract the calls to ensureSnapServicesForGroup from the do* helpers to a single one here

Reply:
so after some analysis, the current code in ensureSnapServicesForGroup is currently idempotent except for one bit where we stop the slice and remove it; most of the idempotency comes from the fact that EnsureSnapServices is itself idempotent. For the bit that is not idempotent, if the slice was already removed then running again that code will do the wrong thing by trying to stop and re-remove, but we can fix that to make that check if the slice exists first, then if it doesn't exist we skip stopping it and removing it and then the whole function is idempotent. 

Now, you say we need to make this whole thing idempotent with tests, do you want that assurance via tests in the same PR or can that come in a followup PR? By my rough estimates its probably another 300+ lines of code to this PR to incorporate those tests here.

Independently I think I will refactor ensureSnapServicesForGroup a bit in a follow-up PR so it can be called instead once from doQuotaControl instead of the individual helpers here but I will do that at the same time that I support multiple quota control actions in one task, which is not a goal for this PR.

I will make the function idempotent for this PR, but I would prefer to leave the tests proving it's idempotency for a follow-up.

Reply:
well actually even the group removal is actually idempotent, as we just log a notice if the stop fails, and the RemoveQuotaGroup function doesn't complain if the file didn't exist, and we even unit test the case where the file doesn't exist, but we should document for EnsureSnapServices and RemoveQuotaGroup that those functions are idempotent.

Reply:
and actually at least on hirsute's systemd version, `systemctl stop` is idempotent for slices that exist and for slices that don't exist, returning 0 in both cases, so actually ignore what I said about it not being fully idempotent, it already is full idempotent.

Reply:
it's the entire operation including the state changes that needs to be idempotent, if that is not, then we need to carefully place the t.SetStatus(state.DoneStatus). The idempotence of ensureSnapServicesForGroup is not enough. The question is what happens if we reboot after one of the unlocks there but before the set status  done.

Reply:
I see your point about the other state changes not being idempotent. They are not currently perfectly idempotent, for example removing a group in the doRemoveQuota handler is not idempotent because the first thing the handler does is ensure that the group exists before continuing, this will fail after we come back from a reboot after unlocking the state since the group will already have been deleted from the state. It is possible to make `doRemoveQuota` and `doCreateQuota` idempotent however by changing the checks that the groups exist and don't exist respectively softer, and if the groups already don't exist or already exist respectively just skip to the `ensureSnapServicesForGroup` part directly. It's a little tricky for the remove case since in the case the group doesn't exist in the state when we re-enter the handler, we would need to create the name and parent chain to get the full slice name which is to be used when stopping and removing the slice, but I think it can be done. The doUpdateQuota handler just needs one slight change to make adding snaps to a group like adding to a set rather than always appending to a list, since currently you would end up with the same snap appended twice in the state which is not what we want, but otherwise the handler should be idempotent. Unifying the update and create cases externally to be the same operation may help with this somewhat, so the most complicated bit is just figuring out how to handle removing quota groups properly. I will have a think about the recreating the necessary group information for the remove case and if that ends up being too complicated then I think we will have to set the done state after `wrappers.EnsureSnapServices` and `wrappers.RemoveQuotaGroup` have run, but before releasing the state lock. Actually now that I say that, it's a bit difficult to figure out if setting the Done status should be before or after the RemoveQuotaGroup...

I wrote the rest of this up before you responded and figured it might be useful to include here anyways, so if you want my analysis on the idempotency of  `ensureSnapServicesForGroup` see below.

# Idempotency of `ensureSnapServicesForGroup`

There are 3 unlocks in `ensureSnapServicesForGroup`:

1. before `wrappers.RemoveQuotaGroup`
2. before `wrappers.QueryDisabledServices`
3. before `wrappers.RestartServices`

## Case 1

If we reboot after unlock 1, then we will have already generated/updated all of the unit files on disk before the unlock, so the second call to `wrappers.EnsureSnapServices()` will do nothing, and the callback will not be called, thus there will be no snaps in `appsToRestartBySnap` and no groups in `grpsToStart`. Thus we will do nothing in lines 478 - 521, and the call to RemoveQuotaGroup will either a) delete the file if we got rebooted before the file was deleted, or b) return nil if the file was already deleted. 

Arguably, we probably could skip unlocking for RemoveQuotaGroup as currently implemented since all that does is delete the file and call daemon-reload. Probably the lock needs to be unlocked instead for `systemd.Stop(grp.SliceFileName(),...)`, and for there the same argument as above holds - if we were rebooted after unlocking but before the slice was stopped, then when we come back after the reboot we will just stop the slice which is now active (since it wasn't deleted before we rebooted), or if we reboot after stopping the slice, then we do the same thing, stopping the slice. If we come back after the slice file was deleted, then systemctl stop does not complain or do anything nebulous so we just continue on as-is.


## Case 2

If we reboot after unlock 2, but before querying the disabled services then nothing changes since this operation doesn't actually mutate any state, it just queries the current state of services. This call could also hypothetically be done with the state lock since it doesn't involve any waiting for services to start or stop. 

## Case 3

This case is arguably the most important one we want to unlock for, since it is the one that could take the most amount of time via start-timeout and stop-timeout in snap.yaml's, but is still safe anyways as follows. If we modified services to be added/removed to slices, then got to this point before the services were restarted and were immediately rebooted, then it is as if the services were restarted anyways since we rebooted the device. Rebooting here is effectively the same as restarting the services, and when we come back since EnsureSnapServices() will do nothing and not call the observer callback function at all, we will not even enter the loop where RestartSnapServices is called at all. 

Similarly, if we were rebooted after the services were unlocked but before the state lock was re-gained, then the services would have caught a rough patch and get restarted twice, but otherwise nothing else is different since they are still all restarted after their unit definitions were modified.

Reply:
as discussed, I removed the state unlocking around the call to wrappers.RemoveQuotaGroup since it's not necessary, and I also added a task argument to the ensureSnapServicesForGroup helper and set the task (if not nil) to be done after calling RemoveQuotaGroup, but before we unlock the state and thus serialize the state changes. I also added a unit test for EnsureSnapAbsentFromQuota to ensure that it is idempotent as expected